### PR TITLE
TMF-F4 get team by uuid

### PR DESCRIPTION
**Pull request by ivshapovalov**
Connect to #11
Connect to #12

### CODE REVIEW

File: src/main/java/ua/com/juja/microservices/teams/dao/TeamRepository.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        Aggregation agg = newAggregation(
+                match(Criteria.where("deactivateDate").gt(actualDate)
+                        .and("members").in(members)),
+                project("members"),
+                unwind("members"),
+                match(Criteria.where("members").in(members)),
+                group("members").count().as("teams"),
+                project("teams").and("name").previousOperation(),
+                sort(Sort.Direction.DESC, "teams")
+        );
```
Comment (OlgaKulikova) [inline src/main/java/ua/com/juja/microservices/teams/dao/TeamRepository.java]: I need explanation what is going on in this agregation.

File: src/main/java/ua/com/juja/microservices/teams/dao/TeamRepository.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    public Team saveTeam(Team team) {
+        log.debug("Started 'Save team' '{}' into DB ", team.toString());
+        mongoTemplate.save(team);
+        log.debug("Finished 'Save team' '{}' into DB ", team.toString());
+        return team;
+    }
+
+    @Getter
+    private class User {
```
Comment (OlgaKulikova) [inline src/main/java/ua/com/juja/microservices/teams/dao/TeamRepository.java]: What is it? Why you need this class? Where do you get user's name?

File: src/main/java/ua/com/juja/microservices/teams/entity/Team.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package ua.com.juja.microservices.teams.entity;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonFormat;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.*;
```
Comment (OlgaKulikova) [inline src/main/java/ua/com/juja/microservices/teams/entity/Team.java]: please don't use * for imports

File: src/main/java/ua/com/juja/microservices/teams/service/TeamService.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    private Team mappingRequestToTeam(TeamRequest teamRequest) {
+        return new Team(teamRequest.getMembers());
+    }
+
+    public Team deactivateTeam(String uuid) {
+        log.debug("Started 'deactivateTeam' with uuid '{}'", uuid);
+        Team team = getUserActiveTeam(uuid);
+        log.debug("Finished 'usersInCurrentTeams' with uuid '{}'. Teams ", uuid, team.toString());
+        team.setDeactivateDate(Date.from(Instant.now()));
```
Comment (OlgaKulikova) [inline src/main/java/ua/com/juja/microservices/teams/service/TeamService.java]: In other places date has such format -  Date.from(LocalDateTime.now().atZone(ZoneId.systemDefault()).toInstant());

File: src/main/java/ua/com/juja/microservices/teams/service/Utils.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.io.IOException;
+import java.io.Reader;
+
+/**
+ * @author Ivan Shapovalov
+ */
+public class Utils {
+
+    public static String convertToString(Reader reader) throws IOException {
```
Comment (OlgaKulikova) [inline src/main/java/ua/com/juja/microservices/teams/service/Utils.java]: If you use this method only in the tests you should replace Utils class to the test folder.

File: src/test/java/ua/com/juja/microservices/acceptance/TeamsAcceptanceTest.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+import org.eclipse.jetty.http.HttpMethod;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import java.io.IOException;
+
+import static ua.com.juja.microservices.teams.service.Utils.convertToString;
```
Comment (OlgaKulikova) [inline src/test/java/ua/com/juja/microservices/acceptance/TeamsAcceptanceTest.java]: please use static imports only for testing libraries (it's common approach). It's difficult to understand quickly where this method from.

File: src/test/java/ua/com/juja/microservices/teams/dao/TeamRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        final Team team1 = new Team(new LinkedHashSet<>(Arrays.asList(userInOneTeam, "user1", "user2", userInSeveralTeams)));
+        final Team team2 = new Team(new LinkedHashSet<>(Arrays.asList(userInSeveralTeams, "user3", "user4", "user5")));
+        final List<Team> expected = new ArrayList<>();
+        expected.add(team1);
+        expected.add(team2);
+
+        List<Team> actual = teamRepository.getUserActiveTeams(userInSeveralTeams, actualDate);
+
+        assertEquals(actual.size(), expected.size());
+        for (int i = 0; i < actual.size(); i++) {
```
Comment (OlgaKulikova) [inline src/test/java/ua/com/juja/microservices/teams/dao/TeamRepositoryTest.java:82]: If the Team entity has equals and hashcode overrided methods, you can check these entities with `assertEquals(expected, actual)` only

File: src/test/java/ua/com/juja/microservices/teams/dao/TeamRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        List<String> actual = teamRepository.checkUsersActiveTeams(members,actualDate);
+
+        assertEquals(actual.size(),2);
+        assertThat(actual, is(expected));
+    }
+
+    @Test
+    @UsingDataSet(locations = "/datasets/getAndDeactivateDataSet.json")
+    public void test_checkUsersActiveTeamsNoOneInSeveralTeamsReturnsEmptyList() {
+        Date actualDate = Date.from(LocalDateTime.now().toInstant(ZoneOffset.UTC));
```
Comment (OlgaKulikova) [inline src/test/java/ua/com/juja/microservices/teams/dao/TeamRepositoryTest.java]: Why you use different approaches to set dates?

File: src/main/java/ua/com/juja/microservices/teams/entity/Team.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    private Date deactivateDate;
+
+    @JsonCreator
+    public Team(@JsonProperty("members") Set<String> members) {
+        this.members = members;
+        this.activateDate = Date.from(LocalDateTime.of(LocalDate.now(), LocalTime.MIN).atZone(ZoneId.systemDefault()).toInstant());
+        this.deactivateDate = Date.from(LocalDateTime.of(LocalDate.now().plusMonths(1).minusDays(1), LocalTime.MAX).atZone(ZoneId.systemDefault()).toInstant());
+    }
+
+    public String toJSON() {
```
Comment (OlgaKulikova) [inline src/main/java/ua/com/juja/microservices/teams/entity/Team.java]: Why we need this method?

**Review by OlgaKulikova — CHANGES_REQUESTED** [review]
Please look at my comments and answer the questions.

File: src/main/java/ua/com/juja/microservices/teams/dao/TeamRepository.java (hunk 2 above)
Comment (ivshapovalov) [reply src/main/java/ua/com/juja/microservices/teams/dao/TeamRepository.java]: This class is used for automapping of Mongo's aggregation results 
        AggregationResults<User> groupResults
                = mongoTemplate.aggregate(agg, Team.class, User.class);
        List<User> usersInActiveTeams = groupResults.getMappedResults();

File: src/main/java/ua/com/juja/microservices/teams/service/TeamService.java (hunk 4 above)
Comment (ivshapovalov) [reply src/main/java/ua/com/juja/microservices/teams/service/TeamService.java]: Fixed

File: src/test/java/ua/com/juja/microservices/teams/dao/TeamRepositoryTest.java (hunk 8 above)
Comment (ivshapovalov) [reply src/test/java/ua/com/juja/microservices/teams/dao/TeamRepositoryTest.java]: Fixed

File: src/test/java/ua/com/juja/microservices/teams/dao/TeamRepositoryTest.java (hunk 7 above)
Comment (OlgaKulikova) [reply src/test/java/ua/com/juja/microservices/teams/dao/TeamRepositoryTest.java:82]: your implementation is correct

File: src/main/java/ua/com/juja/microservices/teams/entity/Team.java (hunk 9 above)
Comment (OlgaKulikova) [reply src/main/java/ua/com/juja/microservices/teams/entity/Team.java]: move it to Utils test class

File: src/main/java/ua/com/juja/microservices/teams/dao/TeamRepository.java (hunk 2 above)
Comment (OlgaKulikova) [reply src/main/java/ua/com/juja/microservices/teams/dao/TeamRepository.java]: Rename to Member with field String uuid.

File: src/main/java/ua/com/juja/microservices/teams/dao/TeamRepository.java (hunk 1 above)
Comment (OlgaKulikova) [reply src/main/java/ua/com/juja/microservices/teams/dao/TeamRepository.java]: Leave it as is (except renaming User to Member and name to uuid)

File: src/main/java/ua/com/juja/microservices/teams/controller/TeamController.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    @GetMapping(produces = "application/json")
+    public ResponseEntity<?> getAllActiveTeams() {
+        //TODO Should be implemented feature TMF-F3
+        return null;
+    }
+
+    @GetMapping(value = "/users/{uuid}", produces = "application/json")
+    public ResponseEntity<?> getTeamByUuid(@PathVariable String uuid) {
+        log.debug("Received 'Get team' request. Get team team of user {}", uuid);
```
Comment (petrokramar) [inline src/main/java/ua/com/juja/microservices/teams/controller/TeamController.java]: Correct this - Get team team

**Review by OlgaKulikova — APPROVED** [review]

**Pull request merged**